### PR TITLE
grpc_transcoder: check response transcoder error

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

To fix https://github.com/envoyproxy/envoy/issues/21490

grpc response transcoder will fail with error if the proto message is too deep (>64),  but grpc_transcoder filter encodeData() function did not check the error status.  This pr adds code to check the error status and adds a test.

Risk Level:  Low
Testing:  Unit-test
Docs Changes: No
Release Notes: Yes

### CODE REVIEW

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 1]
```diff
@@ -656,6 +656,9 @@ Http::FilterDataStatus JsonTranscoderFilter::encodeData(Buffer::Instance& data,
   }
 
   readToBuffer(*transcoder_->ResponseOutput(), data);
+  if (checkIfResponseTranscoderFailed()) {
```
Comment (shuoyang2016) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:659]: remove TODO on line:667?

Comment (qiwzhang) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Done

**Comment by wrowe** [comment]
/assign @lizan 

as code-owner/maintainer

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 2]
```diff
@@ -757,6 +762,25 @@ bool JsonTranscoderFilter::checkIfTranscoderFailed(const std::string& details) {
   return false;
 }
 
+bool JsonTranscoderFilter::checkAndRejectIfResponseTranscoderFailed() {
+  const auto& response_status = transcoder_->ResponseStatus();
+  if (!response_status.ok()) {
+    ENVOY_LOG(debug, "Transcoding response error {}", response_status.ToString());
+    error_ = true;
+    encoder_callbacks_->sendLocalReply(
+        Http::Code::InternalServerError,
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:771]: 502 Bad Gateway is more appropriate for this case?

>    The 502 (Bad Gateway) status code indicates that the server, while
   acting as a gateway or proxy, received an invalid response from an
   inbound server it accessed while attempting to fulfill the request.

**Review by lizan — APPROVED** [review]

**Pull request merged**